### ISSUE TITLE
TestUpdateClusterLayout::test_simple_kill_streaming_node_while_bootstrapping marked as failed with coredump: hang during node shutdown on test teardown

### ISSUE THREAD

**Issue by aleksbykov** [scylladb/scylladb#17246]
Scylla version: 5.5.0~dev-0.20240206.c0c291b98536 with build-id ee177ddf9e48dacd755092d72e31345d862d2dd4

dtest with enabled raft topology: [update_cluster_layout_tests.py::TestUpdateClusterLayout::test_simple_kill_streaming_node_while_bootstrapping[33]](https://jenkins.scylladb.com/job/scylla-staging/job/artsiom_mishuta/job/dtest-pytest-gating/60/artifact/logs-full.release.004/1707317123237_update_cluster_layout_tests.py%3A%3ATestUpdateClusterLayout%3A%3Atest_simple_kill_streaming_node_while_bootstrapping%5B33%5D/) one time failed with coredump, But again in node4 log there is no any traceback info. 

test passed all steps itself.
Logs could be found : [HERE](https://jenkins.scylladb.com/job/scylla-staging/job/artsiom_mishuta/job/dtest-pytest-gating/60/artifact/logs-full.release.004/1707317123237_update_cluster_layout_tests.py%3A%3ATestUpdateClusterLayout%3A%3Atest_simple_kill_streaming_node_while_bootstrapping%5B33%5D/)

after gdb core file next traceback were found:
```
(gdb) thread apply all bt

Thread 4 (LWP 9370):
#0  0x00007fd575b230ea in ?? ()
#1  0x00007fd572bcf7a0 in ?? ()
#2  0x0000000000000008 in ?? ()
#3  0x00007fd572bcf790 in ?? ()
#4  0x0000000000000005 in ?? ()
#5  0x00007fd572bcf7a0 in ?? ()
#6  0x0000000005d65205 in seastar::thread_pool::work (this=this@entry=0x6000001edc00, name=...) at ./seastar/src/core/thread_pool.cc:53
#7  0x0000000005d65513 in seastar::thread_pool::thread_pool(seastar::reactor&, seastar::basic_sstring<char, unsigned int, 15u, true>)::$_0::operator()() const (this=<optimized out>) at ./seastar/src/core/thread_pool.cc:41
#8  std::__invoke_impl<void, seastar::thread_pool::thread_pool(seastar::reactor&, seastar::basic_sstring<char, unsigned int, 15u, true>)::$_0&>(std::__invoke_other, seastar::thread_pool::thread_pool(seastar::reactor&, seastar::basic_sstring<char, unsigned int, 15u, true>)::$_0&) (__f=...) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/invoke.h:61
#9  std::__invoke_r<void, seastar::thread_pool::thread_pool(seastar::reactor&, seastar::basic_sstring<char, unsigned int, 15u, true>)::$_0&>(seastar::thread_pool::thread_pool(seastar::reactor&, seastar::basic_sstring<char, unsigned int, 15u, true>)::$_0&) (__fn=...) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/invoke.h:111
#10 std::_Function_handler<void (), seastar::thread_pool::thread_pool(seastar::reactor&, seastar::basic_sstring<char, unsigned int, 15u, true>)::$_0>::_M_invoke(std::_Any_data const&) (__functor=...) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/std_function.h:290
#11 0x0000000005cdd01b in std::function<void ()>::operator()() const (this=0x5) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/std_function.h:591
#12 seastar::posix_thread::start_routine (arg=0x5) at ./seastar/src/core/posix.cc:90
#13 0x00007fd575aae947 in ?? ()
#14 0x0000000000000000 in ?? ()

Thread 3 (LWP 9371):
#0  0x00007fd575b230ea in ?? ()
#1  0x00007fd5725cf7a0 in ?? ()
#2  0x0000000000000008 in ?? ()
#3  0x00007fd5725cf790 in ?? ()
#4  0x0000000000000009 in ?? ()
#5  0x00007fd5725cf7a0 in ?? ()
#6  0x0000000005d65205 in seastar::thread_pool::work (this=this@entry=0x6010000ff000, name=...) at ./seastar/src/core/thread_pool.cc:53
#7  0x0000000005d65513 in seastar::thread_pool::thread_pool(seastar::reactor&, seastar::basic_sstring<char, unsigned int, 15u, true>)::$_0::operator()() const (this=<optimized out>) at ./seastar/src/core/thread_pool.cc:41
#8  std::__invoke_impl<void, seastar::thread_pool::thread_pool(seastar::reactor&, seastar::basic_sstring<char, unsigned int, 15u, true>)::$_0&>(std::__invoke_other, seastar::thread_pool::thread_pool(seastar::reactor&, seastar::basic_sstring<char, unsigned int, 15u, true>)::$_0&) (__f=...) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/invoke.h:61
#9  std::__invoke_r<void, seastar::thread_pool::thread_pool(seastar::reactor&, seastar::basic_sstring<char, unsigned int, 15u, true>)::$_0&>(seastar::thread_pool::thread_pool(seastar::reactor&, seastar::basic_sstring<char, unsigned int, 15u, true>)::$_0&) (__fn=...) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/invoke.h:111
#10 std::_Function_handler<void (), seastar::thread_pool::thread_pool(seastar::reactor&, seastar::basic_sstring<char, unsigned int, 15u, true>)::$_0>::_M_invoke(std::_Any_data const&) (__functor=...) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/std_function.h:290
#11 0x0000000005cdd01b in std::function<void ()>::operator()() const (this=0x9) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/std_function.h:591
#12 seastar::posix_thread::start_routine (arg=0x9) at ./seastar/src/core/posix.cc:90
#13 0x00007fd575aae947 in ?? ()
#14 0x0000000000000000 in ?? ()

Thread 2 (LWP 9369):
#0  0x00007fd575b2cb4d in ?? ()
#1  0x0000000005d6aff1 in seastar::internal::io_pgetevents (io_context=140554775998464, min_nr=min_nr@entry=1, nr=nr@entry=128, events=events@entry=0x7fd5731ce6d0, timeout=0x0, sigmask=sigmask@entry=0x6010000a6030, force_syscall=<optimized out>) at ./seastar/src/core/linux-aio.cc:163
#2  0x0000000005d66e13 in seastar::reactor_backend_aio::await_events (this=this@entry=0x601000140000, timeout=timeout@entry=-1, active_sigmask=active_sigmask@entry=0x6010000a6030) at ./seastar/src/core/reactor_backend.cc:491
#3  0x0000000005d6753d in seastar::reactor_backend_aio::wait_and_process_events (this=0x601000140000, active_sigmask=0x6010000a6030) at ./seastar/src/core/reactor_backend.cc:558
#4  0x0000000005d1ddfd in seastar::reactor::sleep (this=0x6010000a6000) at ./seastar/src/core/reactor.cc:3403
#5  seastar::reactor::do_run (this=0x6010000a6000) at ./seastar/src/core/reactor.cc:3369
--Type <RET> for more, q to quit, c to continue without paging--
#6  0x0000000005d41881 in seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_0::operator()() const (this=0x6000000edcc0) at ./seastar/src/core/reactor.cc:4554
#7  std::__invoke_impl<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_0&>(std::__invoke_other, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_0&) (__f=...) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/invoke.h:61
#8  std::__invoke_r<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_0&>(seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_0&) (__fn=...) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/invoke.h:111
#9  std::_Function_handler<void (), seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_0>::_M_invoke(std::_Any_data const&) (__functor=...) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/std_function.h:290
#10 0x0000000005cdd01b in std::function<void ()>::operator()() const (this=0x7fd5757eb000) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/std_function.h:591
#11 seastar::posix_thread::start_routine (arg=0x7fd5757eb000) at ./seastar/src/core/posix.cc:90
#12 0x00007fd575aae947 in ?? ()
#13 0x0000000000000000 in ?? ()


Thread 1 (LWP 9203):
#0  0x00007fd575b2cb4d in ?? ()
#1  0x0000000005d6aff1 in seastar::internal::io_pgetevents (io_context=140554782720000, min_nr=min_nr@entry=1, nr=nr@entry=128, events=events@entry=0x7ffcb7f142b0, timeout=0x0, sigmask=sigmask@entry=0x600000296030, force_syscall=<optimized out>) at ./seastar/src/core/linux-aio.cc:163
#2  0x0000000005d66e13 in seastar::reactor_backend_aio::await_events (this=this@entry=0x6000002c0000, timeout=timeout@entry=-1, active_sigmask=active_sigmask@entry=0x600000296030) at ./seastar/src/core/reactor_backend.cc:491
#3  0x0000000005d6753d in seastar::reactor_backend_aio::wait_and_process_events (this=0x6000002c0000, active_sigmask=0x600000296030) at ./seastar/src/core/reactor_backend.cc:558
#4  0x0000000005d1ddfd in seastar::reactor::sleep (this=0x600000296000) at ./seastar/src/core/reactor.cc:3403
#5  seastar::reactor::do_run (this=0x600000296000) at ./seastar/src/core/reactor.cc:3369
#6  0x0000000005d1ce69 in seastar::reactor::run (this=0x7fd575e54000) at ./seastar/src/core/reactor.cc:3200
#7  0x0000000005cabe58 in seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&) (this=this@entry=0x7ffcb7f15f40, ac=ac@entry=33, av=av@entry=0x7ffcb7f18368, func=...) at ./seastar/src/core/app-template.cc:276
#8  0x0000000005cab00d in seastar::app_template::run(int, char**, std::function<seastar::future<int> ()>&&) (this=0x7ffcb7f15f40, ac=33, av=0x7ffcb7f18368, func=...) at ./seastar/src/core/app-template.cc:167
#9  0x0000000001369cba in scylla_main (ac=33, av=0x7ffcb7f18368) at main.cc:672
#10 0x000000000136b711 in std::function<int (int, char**)>::operator()(int, char**) const (this=0x7fd575e54000, this@entry=0x7ffcb7f17df0, __args=__args@entry=0x7ffcb7f18368, __args=__args@entry=0x7ffcb7f18368) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/std_function.h:591
#11 0x000000000136833d in main (ac=33, av=0x7ffcb7f18368) at main.cc:2138
(gdb) 

```

Need to mention that from about 100 runs, this is single which failed with core

**Comment by mykaul** [comment]
This is just stuck on IO threads - anything perhaps in the journal (kernel) ? What kernel version are you using?

**Comment by mykaul** [comment]
Is it reproducible?

**Comment by kbr-scylla** [comment]
@aleksbykov can you provide instructions to reproduce on CI like a job that would run just this test 100 times or more?

**Comment by kbr-scylla** [comment]
> @aleksbykov can you provide instructions to reproduce on CI like a job that would run just this test 100 times or more?

ping

**Comment by aleksbykov** [comment]
@kbr-scylla , i created a CI reproducer: https://jenkins.scylladb.com/job/scylla-staging/job/abykov/job/Dtest/job/master-dtest-with-raft-reproducer/3/ it is running right now. i set to run 100 times single test

**Comment by kbr-scylla** [comment]
Now it looks like it reproduced on every run
it is the test that generates the core:
```
09:03:34,126 793     ccm                            WARNING  cluster.py          :768  | test_simple_kill_streaming_node_while_bootstrapping[1-100]: node4: node4 is still running after 127 seconds. Trying to generate coredump using kill(9796, SIGQUIT)...
09:03:35,127 793     ccm                            WARNING  cluster.py          :768  | test_simple_kill_streaming_node_while_bootstrapping[1-100]: node4: node4 scylla-jmx is still running. Killing process using kill(127, SIGKILL)...
09:03:35,167 793     dtest_setup                    WARNING  dtest_setup.py      :164  | test_simple_kill_streaming_node_while_bootstrapping[1-100]: Moving core file /jenkins/workspace/scylla-staging/abykov/Dtest/master-dtest-with-raft-reproducer/scylla-dtest/scylla.9796.1710320614.core to /jenkins/workspace/scylla-staging/abykov/Dtest/master-dtest-with-raft-reproducer/scylla-dtest/logs-full.release.000/1710320615139_update_cluster_layout_tests.py::TestUpdateClusterLayout::test_simple_kill_streaming_node_while_bootstrapping[1-100]/node4-scylla.9796.1710320614.core
```

the reason being `node4 is still running after 127 seconds`; so I guess it happened during cluster teardown, it was waiting for shutdown and hanged?

**Comment by kbr-scylla** [comment]
Indeed, the message comes from this scylla-ccm code
```python3
    def wait_until_stopped(self, wait_seconds=None, marks=None, dump_core=True):
        """
        Wait until node is stopped after do_stop was called.
          - wait_other_notice: return only when the other live nodes of the
            cluster have marked this node has dead.
          - marks: optional list of (node, mark) to call watch_log_for_death on.
        """
        marks = marks or []
        if wait_seconds is None:
            wait_seconds = 127 if self.scylla_mode() != 'debug' else 600

        start = time.time()
        if self.is_running():
            if not self._wait_until_stopped(wait_seconds):
                if dump_core and self.pid:
                    # Aborting is intended to generate a core dump
                    # so the reason the node didn't stop normally can be studied.
                    self.warning("{} is still running after {} seconds. Trying to generate coredump using kill({}, SIGQUIT)...".format(
                        self.name, wait_seconds, self.pid))
                    try:
                        os.kill(self.pid, signal.SIGQUIT)
```

**Comment by kbr-scylla** [comment]
Last logs from node4:
```
INFO  2024-03-13 09:01:24,110 [shard 0:main] compaction_manager - Asked to stop
INFO  2024-03-13 09:01:24,110 [shard 0:main] view_update_generator - Terminating background fiber
INFO  2024-03-13 09:01:24,110 [shard 0:main] compaction_manager - Stopped
INFO  2024-03-13 09:01:24,110 [shard 1:main] compaction_manager - Asked to stop
INFO  2024-03-13 09:01:24,110 [shard 1:main] view_update_generator - Terminating background fiber
INFO  2024-03-13 09:01:24,110 [shard 1:main] compaction_manager - Stopped
WARN  2024-03-13 09:01:24,110 [shard 0:strm] gossip - failure_detector_loop: Got error in the loop, live_nodes={127.0.17.2, 127.0.17.3, 127.0.17.1}: seastar::sleep_aborted (Sleep is aborted)
INFO  2024-03-13 09:01:24,110 [shard 0:strm] gossip - failure_detector_loop: Finished main loop
INFO  2024-03-13 09:01:24,206 [shard 0:strm] raft - [0315b685-b7a5-4b66-b4be-8f810bb0bb2a] applier fiber stopped because state machine was aborted: seastar::abort_requested_exception (abort requested)
WARN  2024-03-13 09:01:24,749 [shard 0: gms] gossip - === Gossip round FAIL: seastar::abort_requested_exception (abort requested)
INFO  2024-03-13 09:01:24,749 [shard 0: gms] gossip - Gossip loop is not scheduled because it is disabled
INFO  2024-03-13 09:01:26,954 [shard 0:main] raft_group_registry - marking Raft server 20b905f4-e5d3-4e83-beef-f9ddeff328f0 as dead for raft groups
INFO  2024-03-13 09:01:26,966 [shard 0:main] raft_group_registry - marking Raft server f7b6d6c9-6a58-4c5c-b615-469bc5ae61de as dead for raft groups
```

**Comment by kbr-scylla** [comment]
Possible dup: https://github.com/scylladb/scylladb/issues/17608

**Comment by patjed41** [comment]
#17608 is a dup indeed. Shutdown of a bootstrapping node hangs in the same place in both tests.

The culprit is: `auto err = co_await wait_for_topology_request_completion(join_params.request_id);` in `join_token_ring`. The node waits on `co_await _topology_state_machine.event.when();`, but the next event never happens.

Normally during shutdown, `co_await _topology_state_machine.event.when();` would end after making `_topology_state_machine.event` broken in `wait_for_group0_stop`:
```cpp
future<> storage_service::wait_for_group0_stop() {
    _group0_as.request_abort();
    _topology_state_machine.event.broken(make_exception_ptr(abort_requested_exception()));
    co_await when_all(std::move(_raft_state_monitor), std::move(_sstable_cleanup_fiber), std::move(_upgrade_to_topology_coordinator_fiber));
}
```

However, we can't count on `wait_for_group0_stop` on a bootstrapping node. It won't be called if the node is stuck in `join_cluster`. In `join_cluster`, we can rely only on `storage_service._abort_source` and that's what the gossip-based join operation is doing.

The fix for this issue (arguably ugly, but I don't have a better idea yet) is adding such a subscriber at the beginning of `join_cluster`:
```cpp
auto sub = _abort_source.subscribe([this] () noexcept {
    this->_topology_state_machine.event.broken(make_exception_ptr(abort_requested_exception()));
});
```

Another potential problem is that in `join_token_ring` we also have code that uses `_group0_as`, for example `update_topology_with_local_metadata`. I suspect shutdown hangs are also possible there for a similar reason -- `_group0_as` is not triggered on a bootstrapping node. I didn't manage to produce a hang though, so I'm not sure. Anyway, adding `_group0_as.request_abort();` to the subscriber above is an option.

cc @kbr-scylla @gleb-cloudius

**Comment by aleksbykov** [comment]
I got reproduce of this issue in sct also. Sct has nemesis BootstrapStreamingErr. The nemesis stop scylla service during bootstraping of the node after different log messages. and periodically command service stop scylla-service hang, and kill with triggering core dump. If you will need any logs or coredump files. i can provide

**Comment by yarongilor** [comment]
@aleksbykov , @kbr-scylla , looks like it might be reproduced in recent master run - 
```
2024-05-16 09:27:14.579 <2024-05-16 09:27:14.536>: (DatabaseLogEvent Severity.ERROR) period_type=one-time event_id=af8c9154-3e3c-437f-a18f-7db6f422cd0f: type=RUNTIME_ERROR regex=std::runtime_error line_number=2226 node=longevity-100gb-4h-master-db-node-d6d9cb1d-1
2024-05-16T09:27:14.536+00:00 longevity-100gb-4h-master-db-node-d6d9cb1d-1      !ERR | scylla[6895]:  [shard  0:strm] raft_topology - send_raft_topology_cmd(stream_ranges) failed with exception (node state is bootstrapping): std::runtime_error (failed status returned from 0da94314-a4fd-4e19-b5fd-001f200f82ba/10.4.9.156)
```
```
2024-05-16 09:27:14.335 <2024-05-16 09:27:14.221>: (DatabaseLogEvent Severity.ERROR) period_type=one-time event_id=36b7ff97-3e35-4fa5-92fe-5ea140f08eee during_nemesis=BootstrapStreamingError: type=DATABASE_ERROR regex=(^ERROR|!\s*?ERR).*\[shard.*\] line_number=2381 node=longevity-100gb-4h-master-db-node-d6d9cb1d-7
2024-05-16T09:27:14.221+00:00 longevity-100gb-4h-master-db-node-d6d9cb1d-7      !ERR | scylla[6865]:  [shard  0: gms] raft_topology - raft_topology_cmd failed with: seastar::abort_requested_exception (abort requested)
```
in https://argus.scylladb.com/test/bfb2ef96-b741-43ea-9e7d-82e652c0ebdc/runs?additionalRuns[]=d6d9cb1d-7343-4283-b3b6-9e7e9dcf0cc3

**Comment by kbr-scylla** [comment]
@yarongilor why do you think it's this issue? Did you see a hang during shutdown?

**Comment by aleksbykov** [comment]
@yarongilor , i think it is another issue and related to sct.
